### PR TITLE
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa

### PR DESCRIPTION
**Pull request by fcard**
aaaaaaa<sub>aaaaaaa<sub>aaaaaa<sub>aaaaaa...</sub></sub></sub></sub>

### CODE REVIEW

File: base/iostream.jl [hunk 1]
```diff
@@ -321,13 +321,22 @@ function peek(s::IOStream)
     ccall(:ios_peekc, Cint, (Ptr{Void},), s)
 end
 
+
+function _seekback(pos::Integer, s::IOStream)
+    ccall(:ios_seekback, Cint, (Cint, Ptr{Void}), pos, s)
+end
+
+let b1 = Char(0x80), b2 = Char(0x800), b3 = Char(0x10000)
+    global @pure _charbytes(c) = c < b1 ? 1 : c < b2 ? 2 : c < b3 ? 3 : 4
```
Comment (nalimilan) [inline base/iostream.jl]: This function already exists, it's called `codelen` (in `strings/string.jl`).

Comment (fcard) [reply base/iostream.jl]: Heh, I knew something like this had to exist, I spent like two hours looking through all the methods using Char and grepping the files but somehow didn't find this one. Out goes this one then...

File: base/iostream.jl [hunk 2]
```diff
@@ -321,13 +321,18 @@ function peek(s::IOStream)
     ccall(:ios_peekc, Cint, (Ptr{Void},), s)
 end
 
+
+function _seekback(pos::Integer, s::IOStream)
```
Comment (nalimilan) [inline base/iostream.jl]: The convention is that `IO` arguments come first.

File: base/iostream.jl [hunk 3]
```diff
@@ -321,13 +321,18 @@ function peek(s::IOStream)
...
+    ccall(:ios_seekback, Cint, (Cint, Ptr{Void}), pos, s)
+end
+
 function skipchars(io::IOStream, pred; linecomment=nothing)
     while !eof(io)
         c = read(io, Char)
```
Comment (nalimilan) [inline base/iostream.jl]: Can't we use `peekchar` here instead of having to seek back when `!pred(c)`?

Comment (fcard) [reply base/iostream.jl]: That's what the older implementation did, but then you have to do twice the work for every character, so it was slower. I figured a single backtrack at the end of the function would be better than a backtrack for every skipped character.

Comment (nalimilan) [reply base/iostream.jl]: I see, thanks.

File: base/iostream.jl (hunk 2 above)
Comment (fcard) [reply base/iostream.jl]: So it is, I was mimicking my C `ios_seekback` function which in turn was mimicking `ios_ungetc` which I guess was mimicking the standard `ungetc`. I blame Dennis Ritchie.

Should I change my C function as well?

Comment (nalimilan) [reply base/iostream.jl]: I would say so, but I haven't really studied the other functions.

Comment (fcard) [reply base/iostream.jl]: I changed both for consistency. Most of the other `ios` functions do use the io argument first too, so I think that's alright.

**Comment by fcard** [comment]
Thanks @nalimilan for coming over here, working on a PR is considerably nicer when I am not sending emails to 600+ people for everything I do and am not constantly clogging the CI ;P

Travis is doomed to fail because I am too poor to afford more than 50 minutes of his time, I don't even know why I am still running it here. (but `include("test/iostream.jl")` passes and that's good enough for me)

**Comment by fcard** [comment]
Well, let's try some benchmarks:
```julia
using BenchmarkTools

mktemp() do _, file
    println(file, "G")
    flush(file)
    seek(file, 0)
    skipchars(file, islower)
    @benchmark skipchars($file, islower)
end
```
0.5:
```
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     25.848 ns (0.00% GC)
  median time:      26.136 ns (0.00% GC)
  mean time:        33.210 ns (0.00% GC)
  maximum time:     68.385 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     996
  time tolerance:   5.00%
  memory tolerance: 1.00%
```
0.6 (`alpha.201`):
```
BenchmarkTools.Trial: 
  memory estimate:  32 bytes
  allocs estimate:  1
  --------------
  minimum time:     957.476 ns (0.00% GC)
  median time:      1.621 μs (0.00% GC)
  mean time:        1.417 μs (0.00% GC)
  maximum time:     3.419 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     21
  time tolerance:   5.00%
  memory tolerance: 1.00%
```
0.6 (with this):
```
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     28.704 ns (0.00% GC)
  median time:      29.713 ns (0.00% GC)
  mean time:        39.175 ns (0.00% GC)
  maximum time:     87.518 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     992
  time tolerance:   5.00%
  memory tolerance: 1.00%
```
That's a bit better.

---

Let's see that "Large Inputs" benchmark that led to all this trouble:
```julia
using BenchmarkTools

mktemp() do _, file
    # file has to be really large so we don't reach the Inexact error on 0.5
    println(file, ((" "^1000)*"a")^1000000) 
    flush(file)
    seek(file, 0)
    @benchmark (skipchars($file, isspace); read($file, Char))
end
```
0.5:
```
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     20.321 μs (0.00% GC)
  median time:      20.539 μs (0.00% GC)
  mean time:        54.144 μs (0.00% GC)
  maximum time:     171.765 ms (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%
```

0.6:
```
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     12.162 μs (0.00% GC)
  median time:      12.327 μs (0.00% GC)
  mean time:        19.544 μs (0.00% GC)
  maximum time:     31.510 ms (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%
```
If I am doing this right, that's at least a 1.6x improvement more or less. Now we sit down and wonder if it was worth all this trouble... (hey at least I fixed a bug)

File: base/iostream.jl [hunk 4]
```diff
@@ -321,13 +321,18 @@ function peek(s::IOStream)
...
         if c === linecomment
             readline(io)
         elseif !pred(c)
-            seek(io,position(io)-sizeof(string(c)))
+            _seekback(io, codelen(c))
```
Comment (JeffBezanson) [inline base/iostream.jl]: The function `skip` can already do this.

**Review by JeffBezanson — CHANGES_REQUESTED** [review]

File: src/support/ios.c [hunk 5]
```diff
@@ -1083,6 +1083,18 @@ int ios_ungetc(int c, ios_t *s)
     return c;
 }
 
+int ios_seekback(ios_t *s, int pos) {
```
Comment (JeffBezanson) [inline src/support/ios.c]: This isn't necessary. `ios_skip` with a negative argument has a correct implementation of this.

**Comment by JeffBezanson** [comment]
@fcard Don't worry about email notifications --- those of us who get them do so by choice, and they get filtered to separate folders (at least in my case). Similarly don't worry about clogging up CI. It's on us to have enough resources, and if things get too bad (which hasn't been so much of an issue lately) we can just cancel jobs as needed.

File: base/iostream.jl (hunk 4 above)
Comment (fcard) [reply base/iostream.jl]: Well, now don't I feel silly, I was sure I looked through all the IO functions before resorting to this. That's way better, I was feeling guilty for having introduced a C function when I barely use the language (and I probably did it wrong too, didn't I?)

I will use that then.

**Comment by fcard** [comment]
Well, this looks presentable now, thanks @JeffBezanson, @nalimilan! But it's a bit slower now:
```julia
  minimum time:     42.226 ns (0.00% GC)
  median time:      43.480 ns (0.00% GC)
  mean time:        55.865 ns (0.00% GC)
  maximum time:     464.336 ns (0.00% GC)
```
At least now it only takes 2/3 skippable characters for this function to overtake the old one, instead of 40/50. But is there any reason for this? Both `peekchar` and `skip` need to backtrack the iostream, right? Why is peekchar's backtrack more efficient?

**Comment by fcard** [comment]
Oh, I guess peek doesn't need to backtrack at all, reading the source. Still, I assume `skip` does some extra checks that I could avoid, right? What would be a correct `unsafe_skip` implementation that could assume the information that I already have when I use it in my function? Hmm... I need some time to study this whole thing.

Sorry if I am looking silly by over-optimizing this one function, this version might be good enough, but I just really wanted my version to be strictly better than the old one. I don't want to be a `-`, ever, in any sense :(

Well, but there you have it, the one line fix to make this not a "Large" performance regression anymore (should've said "huge", what a fool!)
```diff
-            seek(io,position(io)-sizeof(string(c)))
+            skip(io, -codelen(c))
```
Line 330 of `iostream.jl`, feel free to use it. I am going to keep hitting my head against this until I get it right, or this is going to haunt me forever. You guys can go do more important things now.

Also thanks @JeffBezanson for the reassuring words. It's still going to make me nervous though, but I will work it out eventually. I hope it's not getting too tiring coming up with encouraging words every time I show up to do something :P I always appreciate them, of course, but there is only so much positivity that a community can attempt to give to an innately negative individual before it becomes an annoying shore to try, and I hope we never reach that point.

**Edit**: Also `reason #48` why I didn't want to make a PR right now, I can't keep my mouth shut when I am like this.

**Comment by nalimilan** [comment]
Can you make a PR now then?

**Comment by fcard** [comment]
@nalimilan Is the one-line solution satisfactory despite the small extra overhead? If it's ready to merge then yes, I will make a PR. As long as it's not something that has a chance of staying lingering open for too long I am fine with it.

**Comment by nalimilan** [comment]
At least that's a clear improvement over the current situation. I'm not an expert of this code, but the small performance difference may not be a big issue.

**Comment by fcard** [comment]
Oh, I went to sleep and then wake up to the PR already merged? Cool, thanks! i feel very relieved...

This whole thing was caused because I was testing the code with `IOBuffer`s, but then at the last moment I changed it to `IOStream`s and turns out `seek` isn't nearly as efficient for them... So I panicked and made this whole situation with the C functions and etc. messing with things I clearly didn't understand.

I still am going to try to improve this, studying the `ios_t` structure with a calmer mind to do it properly next time. Even if it doesn't make it to base for being an overcomplicated solution to an insignificant problem, I need it for my peace of mind, and maybe expanding my pool of knowledge to this part of the language will make me actually useful if another issue like this arrives.

I will improve, I swear, I will learn to do things without somehow causing a big mess because I missed a detail and then hurried everyone because I started to panic and nobody had time to actually notice the mistake... I love this language so much, I want to help it grow! And you people have been so nice to me I really don't want to let you down by all this not going anywhere. (I know I shouldn't feel obligated to do this but that already burrowed itself in my brain too deep so no going back...)

I am going to try to improve somewhere else for a while and give you some peace. I hope I am just projecting and my antics aren't as stressful for you as they are for me but just in case they are, I am very very sorry!

That's it for me, lads... Thank you and goodnight.

**Pull request closed without merging**